packages/metrics/src/features/audit/service.ts

```diff
@@ -0,0 +1,122 @@
+ */
+import * as Hapi from '@hapi/hapi'
+
+import { query, writePoints } from '@metrics/influxdb/client'
+import { generateAuditPoint } from '@metrics/features/registration/pointGenerator'
+import {
+  getCompositionIdFromCompositionOrTask,
+  getPractitionerIdFromBundle,
+  getTask,
+  getTrackingId
+} from '@metrics/features/registration/fhirUtils'
+
+type UserAuditAction =
+  | 'DECLARED'
+  | 'REGISTERED'
+  | 'REJECTED'
+  | 'CORRECTED'
+  | 'VALIDATED'
+  | 'ASSIGNED'
+  | 'UNASSIGNED'
+
+type RawUserAuditDataPoint = {
+  practitionerId: number
+  action: UserAuditAction
+  ipAddress: string
+  data: string
+  userAgent: string
+}
+
+type ParsedUserAuditDataPoint = {
+  practitionerId: number
+  action: UserAuditAction
+  ipAddress: string
+  data: Record<string, any>
+  userAgent: string
+}
+
+type UserAuditDataPointWithComposition<Action> = ParsedUserAuditDataPoint & {
+  action: Action
+  data: { compositionId: string }
+}
+
+type UserAuditDataPoint =
+  | UserAuditDataPointWithComposition<'DECLARED'>
+  | UserAuditDataPointWithComposition<'REGISTERED'>
+  | UserAuditDataPointWithComposition<'REJECTED'>
+  | UserAuditDataPointWithComposition<'CORRECTED'>
+  | UserAuditDataPointWithComposition<'VALIDATED'>
+  | UserAuditDataPointWithComposition<'ASSIGNED'>
+  | UserAuditDataPointWithComposition<'UNASSIGNED'>
+
+export async function createUserAuditPointFromFHIR(
+  action: UserAuditAction,
+  request: Hapi.Request
+) {
+  const ipAddress = request.headers['x-real-ip'] || request.info.remoteAddress
+  const userAgent =
+    request.headers['x-real-user-agent'] || request.headers['user-agent']
+
+  const bundle = request.payload as fhir.Bundle
+
+  return writePoints([
+    generateAuditPoint(
+      getPractitionerIdFromBundle(bundle)!,
+      action,
+      ipAddress,
+      userAgent,
+      {
+        compositionId: getCompositionIdFromCompositionOrTask(bundle),
+        trackingId: getTrackingId(getTask(bundle)!)
+      }
+    )
+  ])
+}
+
+export async function getUserAuditEvents(
+  practitionerId: string,
+  size: number,
+  skip: number
+): Promise<Array<UserAuditDataPoint>> {
+  const results = await query<Array<RawUserAuditDataPoint>>(
+    `SELECT * from user_audit_event where practitionerId = $practitionerId limit ${size} offset ${skip}`,
+    {
+      placeholders: { practitionerId: practitionerId }
+    }
+  )
+
+  return results.map((row) => ({
+    ...row,
+    data: row.data ? JSON.parse(row.data) : {}
+  }))
+}
+
+export async function countUserAuditEvents(
+  practitionerId: string
+): Promise<number> {
+  let result
+  try {
+    result = await query(
+      `SELECT COUNT(ipAddress)
+       from user_audit_event
+       where practitionerId = $practitionerId`,
+      { placeholders: { practitionerId: practitionerId } }
+    )
+    return result[0].count
+    // Error will be thrown when
+    // practitioner has no events recorded in influx
+    // return zero in that case
+  } catch (e) {
+    return 0
+  }
```

Comment:
@KaliGreen that comment was super helpful! I think we can also make the code like this so we don't have to throw an error on purpose. I'm thinking if there's a real exception happening for some reason, we're wouldn't just return a 0 and hide the actual error:

```suggestion
  const result = await query(
    `SELECT COUNT(ipAddress)
     from user_audit_event
     where practitionerId = $practitionerId`,
    { placeholders: { practitionerId: practitionerId } }
  )
  if (results.length === 0) {
    return 0
  }
  return result[0].count
```